ambari-server/src/test/java/org/apache/ambari/server/api/services/mpackadvisor/MpackAdvisorHelperTest.java

```diff
@@ -0,0 +1,240 @@
+    doReturn(command).when(helper).createRecommendationCommand(ServiceInfo.ServiceAdvisorType.PYTHON, request);
+    MpackRecommendationResponse response = helper.recommend(request);
+
+    assertEquals(expected, response);
+  }
+
+  @Test(expected = MpackAdvisorException.class)
+  @SuppressWarnings("unchecked")
+  public void testRecommend_commandThrowsException_throwsException() throws MpackAdvisorException,
+      IOException {
+    Configuration configuration = mock(Configuration.class);
+    when(configuration.getRecommendationsArtifactsRolloverMax()).thenReturn(100);
+    MpackAdvisorRunner saRunner = mock(MpackAdvisorRunner.class);
+    AmbariMetaInfo metaInfo = mock(AmbariMetaInfo.class);
+    ServiceInfo service = mock(ServiceInfo.class);
+    when(metaInfo.getService(Mockito.anyString(), Mockito.anyString(), Mockito.anyString())).thenReturn(service);
+    when(service.getServiceAdvisorType()).thenReturn(ServiceInfo.ServiceAdvisorType.PYTHON);
+    MpackAdvisorHelper helper = mpackAdvisorHelperSpy(configuration, saRunner, metaInfo);
+
+    MpackAdvisorCommand<MpackRecommendationResponse> command = mock(MpackAdvisorCommand.class);
+    MpackAdvisorRequest.MpackAdvisorRequestType requestType = MpackAdvisorRequest.MpackAdvisorRequestType.HOST_GROUPS;
+    MpackAdvisorRequest request = MpackAdvisorRequest.MpackAdvisorRequestBuilder.forStack()
+        .ofType(requestType).build();
+
+    when(command.invoke(request)).thenThrow(new MpackAdvisorException("message"));
+    doReturn(command).when(helper).createRecommendationCommand(ServiceInfo.ServiceAdvisorType.PYTHON, request);
+    helper.recommend(request);
+
+    assertTrue(false);
+  }
+
+  @Test
+  public void testCreateRecommendationCommand_returnsComponentLayoutRecommendationCommand()
+      throws IOException, MpackAdvisorException {
+    Configuration configuration = mock(Configuration.class);
+    when(configuration.getRecommendationsArtifactsRolloverMax()).thenReturn(100);
+    MpackAdvisorRunner maRunner = mock(MpackAdvisorRunner.class);
+    AmbariMetaInfo metaInfo = mock(AmbariMetaInfo.class);
+    ServiceInfo service = mock(ServiceInfo.class);
+    when(metaInfo.getService(Mockito.anyString(), Mockito.anyString(), Mockito.anyString())).thenReturn(service);
+    when(service.getServiceAdvisorType()).thenReturn(ServiceInfo.ServiceAdvisorType.PYTHON);
+    MpackAdvisorHelper helper = new MpackAdvisorHelper(configuration, maRunner, metaInfo, null);
+    MpackAdvisorRequest.MpackAdvisorRequestType requestType = MpackAdvisorRequest.MpackAdvisorRequestType.HOST_GROUPS;
+    MpackAdvisorRequest request = MpackAdvisorRequest.MpackAdvisorRequestBuilder.forStack()
+        .ofType(requestType).build();
+
+    MpackAdvisorCommand<MpackRecommendationResponse> command = null;
+    try {
+      command = helper
+          .createRecommendationCommand(ServiceInfo.ServiceAdvisorType.PYTHON, request);
+    } catch (MpackAdvisorException e) {
+      e.printStackTrace();
```

Comment:
Shouldn't catch `MpackAdvisorException`, let the test error with it, instead of `NullPointerException` below.

Reply:
Removed